File: app/src/main/java/com/paper/app/ui/theme/Theme.kt
package com.paper.app.ui.theme

import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Typography
import androidx.compose.material3.darkColorScheme
import androidx.compose.material3.lightColorScheme
import androidx.compose.runtime.Composable
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.sp

// Minimalist, paper-like: near-black ink on warm off-white, no accent noise.
private val Ink = Color(0xFF1A1A1A)
private val PaperWhite = Color(0xFFFAF8F5)
private val FaintLine = Color(0xFFE5E1DB)
private val SoftGray = Color(0xFF8A857E)

private val NightInk = Color(0xFFE8E6E3)
private val NightPaper = Color(0xFF141414)
private val NightLine = Color(0xFF2A2A2A)

private val LightColors = lightColorScheme(
    primary = Ink,
    onPrimary = PaperWhite,
    background = PaperWhite,
    onBackground = Ink,
    surface = PaperWhite,
    onSurface = Ink,
    surfaceVariant = FaintLine,
    onSurfaceVariant = SoftGray,
    outline = FaintLine
)

private val DarkColors = darkColorScheme(
    primary = NightInk,
    onPrimary = NightPaper,
    background = NightPaper,
    onBackground = NightInk,
    surface = NightPaper,
    onSurface = NightInk,
    surfaceVariant = NightLine,
    onSurfaceVariant = SoftGray,
    outline = NightLine
)

private val PaperTypography = Typography(
    headlineMedium = TextStyle(
        fontFamily = FontFamily.Serif,
        fontWeight = FontWeight.Normal,
        fontSize = 28.sp
    ),
    titleMedium = TextStyle(
        fontFamily = FontFamily.Serif,
        fontWeight = FontWeight.Medium,
        fontSize = 18.sp
    ),
    bodyLarge = TextStyle(
        fontFamily = FontFamily.Default,
        fontSize = 16.sp,
        lineHeight = 26.sp
    ),
    bodyMedium = TextStyle(
        fontFamily = FontFamily.Default,
        fontSize = 14.sp,
        lineHeight = 22.sp
    ),
    labelMedium = TextStyle(
        fontFamily = FontFamily.Default,
        fontWeight = FontWeight.Medium,
        fontSize = 13.sp,
        letterSpacing = 0.5.sp
    )
)

@Composable
fun PaperTheme(content: @Composable () -> Unit) {
    MaterialTheme(
        colorScheme = if (isSystemInDarkTheme()) DarkColors else LightColors,
        typography = PaperTypography,
        content = content
    )
}

File: PaperiOS/Sources/UI/Theme.swift
import SwiftUI

/// Minimalist, paper-like palette ported verbatim from the Android theme
/// (`ui/theme/Theme.kt`): near-black ink on warm off-white, no accent noise.
enum PaperPalette {
    // Light
    static let ink = Color(hex: 0x1A1A1A)
    static let paperWhite = Color(hex: 0xFAF8F5)
    static let faintLine = Color(hex: 0xE5E1DB)
    static let softGray = Color(hex: 0x8A857E)
    // Dark
    static let nightInk = Color(hex: 0xE8E6E3)
    static let nightPaper = Color(hex: 0x141414)
    static let nightLine = Color(hex: 0x2A2A2A)
}

/// Semantic colors resolved against the active color scheme — mirrors the
/// Material `colorScheme` roles the Android screens read.
struct PaperColors {
    let scheme: ColorScheme

    var background: Color { scheme == .dark ? PaperPalette.nightPaper : PaperPalette.paperWhite }
    var onBackground: Color { scheme == .dark ? PaperPalette.nightInk : PaperPalette.ink }
    var onSurfaceVariant: Color { PaperPalette.softGray }
    var outline: Color { scheme == .dark ? PaperPalette.nightLine : PaperPalette.faintLine }
    /// Text/icon color on top of the inked primary button.
    var onPrimary: Color { scheme == .dark ? PaperPalette.nightPaper : PaperPalette.paperWhite }
    var primary: Color { onBackground }
}

/// Persists the light/dark override (nil = follow system), matching the Android
/// `ThemePreference` toggle behavior.
final class ThemeManager: ObservableObject {
    private let key = "dark_mode"
    private let defaults: UserDefaults

    /// nil = follow system; true/false = explicit user override.
    @Published var darkModeOverride: Bool? {
        didSet {
            if let v = darkModeOverride {
                defaults.set(v, forKey: key)
            } else {
                defaults.removeObject(forKey: key)
            }
        }
    }

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        self.darkModeOverride = defaults.object(forKey: key) as? Bool
    }

    /// SwiftUI `preferredColorScheme` value (nil follows the system).
    var colorScheme: ColorScheme? {
        guard let v = darkModeOverride else { return nil }
        return v ? .dark : .light
    }

    /// Flips relative to what's currently showing, then persists the explicit choice.
    func toggle(current: ColorScheme) {
        let isDark = darkModeOverride ?? (current == .dark)
        darkModeOverride = !isDark
    }
}

extension Font {
    /// Serif for headlines/titles, matching Android's `FontFamily.Serif`.
    static func paperSerif(_ size: CGFloat, weight: Font.Weight = .regular) -> Font {
        .system(size: size, weight: weight, design: .serif)
    }
}

extension Color {
    init(hex: UInt32) {
        self.init(
            red: Double((hex >> 16) & 0xFF) / 255.0,
            green: Double((hex >> 8) & 0xFF) / 255.0,
            blue: Double(hex & 0xFF) / 255.0
        )
    }
}
